lib/src/arguments.dart

```diff
@@ -0,0 +1,198 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// --------------------------------------------------
+//
+// This lsif_indexer software is based on a number of software repositories with
+// separate copyright notices and/or license terms. Your use of the source
+// code for the these software repositories is subject to the terms and
+// conditions of the following licenses:
+//
+// lsif-dart: https://github.com/sourcegraph/lsif-dart
+// Copyright Anton Astashov. All rights reserved.
+// Licensed under the BSD-2 Clause License: https://github.com/sourcegraph/lsif-dart/blob/master/LICENSE
+//
+// crossdart: https://github.com/astashov/crossdart
+// Copyright Anton Astashov. All rights reserved.
+// Licensed under the BSD-2 Clause License: https://github.com/astashov/crossdart/blob/master/LICENSE
+
+import 'package:args/args.dart';
+import 'package:meta/meta.dart';
+
+/// The resulting config of parsing arguments for the lsif_indexer program.
+class LsifDartConfig {
+  /// All arguments that are either options (eg. -o file.ext) or flags (eg. -h).
+  final Map<String, dynamic> options;
+
+  /// All arguments provided after [options].
+  final List<String> rest;
+
+  /// If the command line arguments are valid.
+  final bool isValid;
+
+  const LsifDartConfig(this.options, this.rest, {this.isValid = true});
+
+  /// The provided output destination to store the LSIF results in. Can be `null`.
+  String get output => _valueForItem(Config.output);
+
+  /// The provided project root for analysis. Can be `null`.
+  String get projectRoot => _valueForItem(Config.projectRoot);
+
+  String _valueForItem(_ConfigItem item) => options[item.name];
+}
+
+class LsifDartArgumentParser {
+  final ArgParser _argParser;
+
+  LsifDartArgumentParser() : _argParser = ArgParser() {
+    _argParser.addItems(Config.all);
+  }
+
+  LsifDartConfig parse(List<String> arguments) {
+    final argResult = _argParser.parse(arguments);
+
+    if (argResult['help'] == true) {
+      _showHelp();
+      return LsifDartConfig({}, [], isValid: false);
+    }
+
+    final options = argResult.options.fold(
+      <String, dynamic>{},
+      (accumulator, option) {
+        accumulator[option] = argResult[option];
+        return accumulator;
+      },
+    );
+
+    return LsifDartConfig(options, argResult.rest);
+  }
+
+  void _showHelp([String errorMessage]) {
+    print(
+      'LSIF Dart Indexer - A Dart language server index format generator.\n',
+    );
+    print(
+      'lsif_indexer.dart analyzes all the files of the given project, '
+      'and stores the analyze information in the specified file.',
```

Comment:
nit: analyzed, or analysis information. And maybe should say something like writes to LSIF and have a link to what LSIF is.